### PR TITLE
Transaction Interfaces in meta_state_service

### PR DESCRIPTION
**Pull request by shengofsun**
1. Adjustment of transaction interface according to pr #155, resolve issue #146 
2. Implement transaction interface for meta_state_service_zookeeper and meta_state_service_simple, and add tests for them

### CODE REVIEW

File: include/dsn/dist/error_code.h [hunk 1]
```diff
@@ -42,3 +42,5 @@ DEFINE_ERR_CODE(ERR_HOLD_BY_OTHERS)
 DEFINE_ERR_CODE(ERR_RECURSIVE_LOCK)
 DEFINE_ERR_CODE(ERR_NO_OWNER)
 DEFINE_ERR_CODE(ERR_NODE_ALREADY_EXIST)
+DEFINE_ERR_CODE(ERR_CONSISTENCY)
```
Comment (imzhenyu) [inline include/dsn/dist/error_code.h]: ERR_INCONSISTENT_STATE?

**Comment by imzhenyu** [comment]
for the unit test cases, can we write them in a way that they are provider independent? There are several benefits by doing this:  (1) one case for all and future providers to avoid redundant work; (2) make sure the usages are only replying on the provider interface.

File: include/dsn/dist/error_code.h (hunk 1 above)
Comment (shengofsun) [inline include/dsn/dist/error_code.h]: You are right, this is more clear

**Comment by shengofsun** [comment]
Meta_state_service can use the same code for UT, and I can't agree with you any more on this. But it's not this case for the distributed_lock part, because distributed_lock_service_simple don't support the multiple apps to acquire the same lock at the same time.

**Comment by imzhenyu** [comment]
great, then I will wait for your refactoring for sharing the UT for the two different meta state providers.

**Comment by shengofsun** [comment]
@imzhenyu refactor of UT done

**Comment by imzhenyu** [comment]
seems good to me. how is the status of multi-meta server fail-over test with zookeeper now?

**Comment by shengofsun** [comment]
I tried to kill meta manually, it worked. Later I will kill meta periodically with a script.

**Comment by imzhenyu** [comment]
Cool~ Then we can setup ZK on windows as well:-)

**Pull request merged**